src/Microsoft.Diagnostics.Monitoring.EventPipe/Logs/EventLogsPipeline.cs

```diff
@@ -42,44 +42,48 @@ protected override Task OnEventSourceAvailable(EventPipeEventSource eventSource,
-            Stack<Guid> stack = new();
+            //
+            // We enable TplEventSource's TasksFlowActivityIds as part of our configuration to enable activity correlation.
+            // This means that each time an event start occurs the current ActivityId will branch creating a new one with a RelatedActivityId equal to where it branched from.
+            // Combining this with the fact that scopes are handled as ActivityJson/{Start,Stop} means the ActivityId will branch each time a scope starts.
+            // When a log message occurs, it'll have an ActivityId equal to the latest applicable scope.
+            //
+            // By maintaining a tree with the branching data, we can construct the full scope for a log message:
+            // - Each time the ActivityId branches, create a node in the tree with it's parent being the node corresponding to the RelatedActivityId.
+            //   - Each node has corresponding scope data.
+            // - When a log message occurs, grab the node with the corresponding ActivityId and backtrack to the root of the tree. Each node visited is included as part of the log's scope.
+            //
+            // NOTE: There are edge cases with concurrent traces, this is described in greater detail above our backtracking code.
+            //
+            Dictionary<Guid, LogScopeItem> activityIdToScope = new();
 
             eventSource.Dynamic.AddCallbackForProviderEvent(LoggingSourceConfiguration.MicrosoftExtensionsLoggingProviderName, "ActivityJson/Start", (traceEvent) => {
-                int factoryId = (int)traceEvent.PayloadByName("FactoryID");
-                string categoryName = (string)traceEvent.PayloadByName("LoggerName");
+                if (traceEvent.ActivityID == Guid.Empty)
+                {
+                    // Unexpected
+                    return;
+                }
+
                 string argsJson = (string)traceEvent.PayloadByName("ArgumentsJson");
 
                 // TODO: Store this information by logger factory id
-                LogActivityItem item = new()
+                LogScopeItem item = new()
                 {
                     ActivityID = traceEvent.ActivityID,
                     ScopedObject = new LogObject(JsonDocument.Parse(argsJson).RootElement),
                 };
 
-                if (stack.Count > 0)
+                if (activityIdToScope.TryGetValue(traceEvent.RelatedActivityID, out LogScopeItem parentItem))
                 {
-                    Guid parentId = stack.Peek();
-                    if (logActivities.TryGetValue(parentId, out LogActivityItem parentItem))
-                    {
-                        item.Parent = parentItem;
-                    }
+                    item.Parent = parentItem;
                 }
 
-                stack.Push(traceEvent.ActivityID);
-                logActivities[traceEvent.ActivityID] = item;
+                activityIdToScope[traceEvent.ActivityID] = item;
```

Comment:
As a safety precaution you might want to set an upper bound on the size of this dictionary, say 10_000 entries and then you ignore any further scopes? This protects against buggy or pathological apps turning into out-of-memory for dotnet-monitor. I don't think it is guaranteed that a scope will be disposed with the same activity ID active as when it was created, but all the examples I can think of hopefully the dev would notice because their scopes are leaking or they'd have to  do something pretty funky in their code. I hope it will be very rare.